File: src/main/java/site/yapp/study/todolist/api/todo/controller/TodoController.java
package site.yapp.study.todolist.api.todo.controller;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import site.yapp.study.todolist.api.todo.dto.request.TodoBulkCreateRequestDto;
import site.yapp.study.todolist.api.todo.dto.request.TodoCreateRequestDto;
import site.yapp.study.todolist.api.todo.dto.request.TodoUpdateRequestDto;
import site.yapp.study.todolist.api.todo.dto.response.TodoGetResponseDto;
import site.yapp.study.todolist.api.todo.dto.response.TodoToggleGetResponseDto;
import site.yapp.study.todolist.api.todo.service.TodoService;
import site.yapp.study.todolist.common.response.ApiResponse;
import site.yapp.study.todolist.common.response.SuccessCode;

import java.util.List;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/v1/todos")
@Tag(name = "[Todo] 투두리스트 관련 API (V1)")
public class TodoController {
    private final TodoService todoService;

    @Operation(summary = "할일 생성 API")
    @ApiResponses(value = {
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "200", description = "할일 생성 성공"),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "400", description = "할일 생성 실패", content = @Content),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "500", description = "서버 내부 오류", content = @Content)
    })
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public ApiResponse<Object> createTodo(@RequestBody TodoCreateRequestDto todoCreateRequestDto) {
        todoService.createTodo(todoCreateRequestDto);
        return ApiResponse.success(SuccessCode.CREATE_TODO_SUCCESS);
    }

    @Operation(summary = "벌크 할일 생성 API")
    @ApiResponses(value = {
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "200", description = "벌크 할일 생성 성공"),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "400", description = "벌크 할일 생성 실패", content = @Content),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "500", description = "서버 내부 오류", content = @Content)
    })
    @PostMapping("/bulk")
    @ResponseStatus(HttpStatus.CREATED)
    public ApiResponse<Object> createBulkTodo(@RequestBody TodoBulkCreateRequestDto todoBulkCreateRequestDto) {
        todoService.createBulkTodo(todoBulkCreateRequestDto);
        return ApiResponse.success(SuccessCode.CREATE_BULK_TODO_SUCCESS);
    }

    @Operation(summary = "할일 전체 조회 API")
    @ApiResponses(value = {
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "200", description = "할일 전체 조회 성공"),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "400", description = "할일 전체 조회 실패", content = @Content),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "500", description = "서버 내부 오류", content = @Content)
    })
    @GetMapping
    @ResponseStatus(HttpStatus.OK)
    public ApiResponse<List<TodoGetResponseDto>> getAllTodo() {
        List<TodoGetResponseDto> response = todoService.getAllTodo();
        return ApiResponse.success(SuccessCode.GET_TODO_SUCCESS, response);
    }

    @Operation(summary = "할일 개별 조회 API")
    @ApiResponses(value = {
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "200", description = "할일 개별 조회 성공"),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "400", description = "할일 개별 조회 실패", content = @Content),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "404", description = "존재하지 않는 할일", content = @Content),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "500", description = "서버 내부 오류", content = @Content)
    })
    @GetMapping("/{todoId}")
    @ResponseStatus(HttpStatus.OK)
    public ApiResponse<TodoGetResponseDto> getEachTodo(@PathVariable Long todoId) {
        TodoGetResponseDto response = todoService.getEachTodo(todoId);
        return ApiResponse.success(SuccessCode.GET_EACH_TODO_SUCCESS, response);
    }

    @Operation(summary = "할일 수정 API")
    @ApiResponses(value = {
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "200", description = "할일 수정 성공"),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "400", description = "할일 수정 실패", content = @Content),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "404", description = "존재하지 않는 할일", content = @Content),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "500", description = "서버 내부 오류", content = @Content)
    })
    @PatchMapping("/{todoId}")
    public ApiResponse<Object> updateTodo(@PathVariable Long todoId, @RequestBody TodoUpdateRequestDto todoUpdateRequestDto) {
        todoService.updateTodo(todoId, todoUpdateRequestDto);
        return ApiResponse.success(SuccessCode.UPDATE_TODO_SUCCESS);
    }

    @Operation(summary = "할일 삭제 API")
    @ApiResponses(value = {
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "200", description = "할일 삭제 성공"),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "400", description = "할일 삭제 실패", content = @Content),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "404", description = "존재하지 않는 할일", content = @Content),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "500", description = "서버 내부 오류", content = @Content)
    })
    @DeleteMapping("/{todoId}")
    public ApiResponse<Object> deleteTodo(@PathVariable Long todoId) {
        todoService.deleteTodo(todoId);
        return ApiResponse.success(SuccessCode.DELETE_TODO_SUCCESS);
    }

    @Operation(summary = "할일 완료 토글 변경 API")
    @ApiResponses(value = {
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "200", description = "할일 완료 토글 변경 성공"),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "400", description = "할일 완료 토글 변경 실패", content = @Content),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "404", description = "존재하지 않는 할일", content = @Content),
            @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "500", description = "서버 내부 오류", content = @Content)
    })
    @PatchMapping("/toggle/{todoId}")
    public ApiResponse<TodoToggleGetResponseDto> toggleTodoStatus(@PathVariable Long todoId, @RequestParam Boolean isCompleted) {
        TodoToggleGetResponseDto response = todoService.toggleTodoStatus(todoId, isCompleted);
        return ApiResponse.success(SuccessCode.TOGGLE_TODO_SUCCESS, response);
    }
}

File: src/main/kotlin/com/example/studytodolist/todo/controller/TodoController.kt
package com.example.studytodolist.todo.controller

import com.example.studytodolist.todo.dto.request.BulkSaveRequestDto
import com.example.studytodolist.todo.dto.request.TodoSaveRequestDto
import com.example.studytodolist.todo.dto.request.TodoUpdateRequestDto
import com.example.studytodolist.todo.dto.response.BulkSaveResponseDto
import com.example.studytodolist.todo.dto.response.TodoFindResponseDto
import com.example.studytodolist.todo.dto.response.TodoSaveResponseDto
import com.example.studytodolist.todo.dto.response.TodoUpdateResponseDto
import com.example.studytodolist.todo.service.TodoService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.net.URI

@Tag(name = "Todo v1.0 API", description = "TodoList 관련 api")
@RestController
@RequestMapping(value = ["/api/v1/todo"], produces = [MediaType.APPLICATION_JSON_VALUE])
class TodoController(private val todoService: TodoService) {
    @Operation(
        summary = "Todo 생성",
        description = "Todo 정보를 받아 생성한다.")
    @PostMapping("")
    fun save(@RequestBody saveRequest: TodoSaveRequestDto): ResponseEntity<TodoSaveResponseDto> {
        val todoSaveResponseDto = todoService.save(saveRequest)
        return ResponseEntity.created(URI.create("/api/v1/todo/".plus(todoSaveResponseDto.id.toString())))
            .body(todoSaveResponseDto)
    }

    @Operation(
        summary = "Todo 리스트 조회",
        description = "모든 Todo를 반환한다.")
    @GetMapping("")
    fun findAll(): ResponseEntity<List<TodoFindResponseDto>> = ResponseEntity.ok(todoService.findAll())

    @Operation(
        summary = "id로 Todo 조회",
        description = "id를 받아 해당하는 Todo를 반환한다.")
    @GetMapping("/{id}")
    fun findById(@PathVariable id: Long): ResponseEntity<TodoFindResponseDto> = ResponseEntity.ok(todoService.findById(id))

    @Operation(
        summary = "Todo 삭제",
        description = "id를 받아 해당하는 Todo를 삭제한다.")
    @DeleteMapping("")
    fun delete(@RequestParam id: Long): ResponseEntity<Void>{
        todoService.deleteById(id)
        return ResponseEntity.noContent().build()
    }

    @Operation(
        summary = "Todo progress 업데이트",
        description = "id와 progress를 받아 해당하는 Todo의 progress를 업데이트한다.")
    @PatchMapping("")
    fun update(@RequestBody updateRequest: TodoUpdateRequestDto): ResponseEntity<TodoUpdateResponseDto> =
        ResponseEntity.ok(todoService.update(updateRequest))

    @Operation(
        summary = "Todo bulk 생성",
        description = "count를 받아 해당하는 개수의 mock data를 bulk save 한 후, 생성한 개수를 반환한다.")
    @PostMapping("/bulk")
    fun bulkSave(@RequestBody bulkSaveRequestDto: BulkSaveRequestDto): ResponseEntity<BulkSaveResponseDto>{
        return ResponseEntity.created(URI.create("/api/v1/todo")).body(todoService.bulkSave(bulkSaveRequestDto))
    }
}